text/XXX-unwind-FFI.md

```diff
@@ -0,0 +1,286 @@
+to unwind the C stack frames.
+
+However, if the C code is compiled
+with support for the same stack-unwinding mechanism
+used to by the Rust code,
+unwinding across the FFI boundary is well-defined and safe,
+and in fact it can be a useful way to handle errors
+when working with certain C libraries, such as `libjpeg`.
+
+Thus, the `may_panic` function may be makred `#[unwind(allow)]`,
+which ensures that the unwinding operation will be propagated to the caller
+rather than aborting the process.
+This annotation will also prevent the compiler
+from marking the function as `nounwind`
+(which permits the backend toolchain from optimizing based on the assumption
+that an unwinding operation cannot escape from a function).
+
+This annotation can only be used with an `unsafe` function,
+since the compiler is unable to guarantee
+that the caller will be compiled with a compatible stack-unwinding mechanism.
+
+This RFC does, however, provide a tool that will enable users
+to provide this guarantee themselves.
+We will stabilize the `!#[panic_runtime]` annotation,
+which [designates a crate as the provider of the final product's panic runtime]
+(less-unwinding).
+Additionally, the standard library's current `panic=unwind` runtime crate,
+`libpanic_unwind`, which is compatible with native C++ style exceptions,
+will be provided under another name (such as `libpanic_native`)
+that guarantees the implementation will remain compatible with C++ exceptions
+(while the implementation of `libpanic_unwind` itself may change
+to no longer maintain that compatibility).
+
+In order for Cargo users to be able
+to specify the C++ compatible `panic_runtime` implementation,
+a new optional value, runtime, will be added to the `profile.foo.panic` option:
+
+```toml
+[profile.dev]
+panic = { strategy = 'unwind', runtime = 'native' }
+```
+
+`panic.runtime` may only be specified if the panic strategy is `unwind`.
+And just as there may only be one strategy
+in the dependency graph for a final product,
+there may only be one `runtime`.
+
+For now, the only valid `runtime` keys will be:
+
+* `default` - identical to `panic = 'unwind'` with no `runtime` selected.
+  This will use Rust's default unwinding runtime,
+  unless another crate in the final product's dependency graph
+specifies `panic = 'abort'`,
+  in which case the `abort` strategy will take precedence as usual.
+  which is not guaranteed to be compatible with native exceptions.
+* `native` - as discussed above, this will preserve the behavior of the current
+  implementation of `libpanic_unwind`.
+* `crate` - indicates that a non-`std` crate will use `#![panic_runtime]` to
+  provide the runtime.
+* `self` - indicates that this crate itself provides `#![panic_runtime]`.
+
+If the `native` or `crate` key is specified
+anywhere in a final product's dependency graph,
+no crate in that dependency graph may specify the `panic = abort` strategy;
+this mismatch will cause the build to fail.
```

Comment:
This lacks motivation. `abort` is strictly more compatible than any implementation of unwinding. `abort` also matches the intended semantics of unwinding in Rust.

Reply:
Thinking about your "intended semantics of unwinding" comment has led me to rethink the motivation of the whole RFC.

In another Internals thread, I [wrote](https://users.rust-lang.org/t/porting-mozjpeg-to-rust-10k-gig/28899/12?u=batmanaod),

> I think Rust should be able to interoperate with existing libraries that use standard non-local control flow constructs...

So, to me, the fundamental semantic mismatch is that C++ treats stack-unwinding as a control-flow construct, while Rust treats it as a way of handling bugs. I think Rust's philosophy on this is _better_, but I stand by the implied statement in my quote above that C++'s view is "standard". (In fact, I believe I've read a Ruby book that explicitly explains exceptions in terms of "nonlocal control flow" rather than in terms of error handling.)

So, to me, the fundamental questions that must be answered are:

* Should Rust have a means of interfacing with foreign APIs that use stack-unwinding as a means of (semantically) non-exceptional control flow?
* If so, how?

I would be okay with putting this RFC PR on hold in order to go completely back to the drawing board on these questions, but I really feel like I need more help than I'm getting currently; I don't know nearly enough about how stack unwinding is actually _implemented_ (and some of the other issues you've pointed in the RFC out reveal this ignorance), and unfortunately I'm not sure I can commit the time to learning these details in any sort of timely manner. (At my current job nearly all our code runs on the JVM, so my work on this PR is in no way relevant to my "real" work.)

Reply:
If we do go back to the drawing board, here are two thoughts I would like to take the time to consider in more depth:

* In multi-threaded code, `panic` *is* actually semantically more like `throw` than in single-threaded code, since by default it only terminates one thread, and the "exception" information (i.e. the `Error` returned from `join`) may be useful to the parent thread. Perhaps it would be good to extend this model, permitting native exceptions to be caught by `join` but not by `catch_panic` (i.e. effectively limiting FFI-stack-unwinding to child threads).
* The only other non-local control flow construct I can think of in Rust that permits returning to a prior stack frame is `await`. Maybe the most idiomatic (albeit nonstandard compared to existing languages) way to deal with exceptions-as-control-flow would be to model FFI code that unwinds as something akin to a generator that may `yield` an error rather than a valid value. (I realize this idea is at best half-baked.)